### PR TITLE
test: migrate processing batch 1 to JUnit 5

### LINKED ISSUE #13948: JUnit4 to JUnit5 migration
Hi team,

A proposal to moving on after the adoption of java8+ is to start the migration from junit4 to jnuit5.

- JUnit 4 never advanced beyond Java 7, missing out on a lot of features from Java 8. JUnit 5 makes good use of the Java 8 features.
- A single jar library contains the entire framework. We need to import the whole library, even when we only require a particular feature. In JUnit 5, we get more granularity and can import only what's necessary.
- Only one test runner can execute tests at a time in JUnit 4 (e.g. @RunWith(Parameterized.class)
 or provide druid extension with the adoption of JUnit Platform). JUnit 5 allows multiple runners to work simultaneously.

As a strategy for the approach I would suggest to migrate one module at a time(it will be easier for the prs) and create tasks for the rest of the modules. As a starting point we can still depend on JUnit Vintage for backward compatibility with JUnit 4 but gradually moving to JUnit5.

I think all the point mentioned describes why it would be beneficial for apache druid as well as I would like to pick this issue if the rest of the community is interested.

### PR DESCRIPTION
**Pull request by FrankChen021**
Fixes #13948.

### Description

This PR migrates processing test Batch 1 to JUnit 5 following `dev/junit5-migration-guidance.md`.

#### Migrate processing tests

- Convert the Batch 1 processing tests to JUnit 5 lifecycle, assertions, and extensions.

#### Add and document JUnit 5 helpers

- Add `TemporaryFolderExtension` for Druid-managed temporary directories.
- Add `LoggerCaptureExtension` for capturing Log4j events in JUnit 5 tests.
- Retain `ExpectThrows` for existing annotation-style tests while preferring `Assertions.assertThrows(...)` for new tests.
- Document the helper APIs and their intended usage in `AGENTS.md`.

The temporary-folder compatibility bridge is intentionally limited to existing mixed JUnit 4/JUnit 5 fixture consumers and can be removed after those callers are migrated.

#### Release note

No user-facing behavior change.

##### Key changed/added classes in this PR

 * `TemporaryFolderExtension`
 * `LoggerCaptureExtension`
 * `ExpectThrows`
 * `NestedDataTestUtils`

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified test helpers.
- [x] added Javadocs for the new helper classes and compatibility bridge.
- [x] added comments explaining the intent of the compatibility bridge.
- [x] added or updated unit tests through the JUnit 5 migration.

Validation:

- `mvn -ntp -pl processing test-compile -Dweb.console.skip=true -T1C`
- Focused processing tests covering the migrated assertion and temporary-folder helpers: 122 tests passed.
- `git diff --check`

### CODE REVIEW

File: processing/src/test/java/org/apache/druid/java/util/common/StringUtilsTest.java [hunk 1]
```diff
@@ -150,165 +145,167 @@
...
   }
 
   @SuppressWarnings("MalformedFormatString")
   @Test
   public void testNonStrictFormat()
   {
-    Assert.assertEquals("test%d; format", StringUtils.nonStrictFormat("test%d", "format"));
-    Assert.assertEquals("test%s%s; format", StringUtils.nonStrictFormat("test%s%s", "format"));
+    Assertions.assertEquals("test%d; format", StringUtils.nonStrictFormat("test%d", "format"));
+    Assertions.assertEquals("test%s%s; format", StringUtils.nonStrictFormat("test%s%s", "format"));
```
[the remark this answers is not in the document]
Comment (FrankChen021) [reply processing/src/test/java/org/apache/druid/java/util/common/StringUtilsTest.java]: Addressed in commit 166af79795 with the repository CodeQL suppression marker for this intentional nonStrictFormat test case; the test behavior is unchanged. The focused processing tests and validation passed.

File: processing/src/test/java/org/apache/druid/java/util/common/CompressionUtilsTest.java [hunk 2]
```diff
@@ -610,49 +610,49 @@
...
+      Assertions.assertEquals(
           GZ_BYTES.length,
           ByteStreams.copy(
               new ZeroRemainingInputStream(new ByteArrayInputStream(GZ_BYTES)),
               outputStream
           )
       );
     }
-    Assert.assertEquals(GZ_BYTES.length * 3, testFile.length());
+    Assertions.assertEquals(GZ_BYTES.length * 3, testFile.length());
```
[the remark this answers is not in the document]
Comment (FrankChen021) [reply processing/src/test/java/org/apache/druid/java/util/common/CompressionUtilsTest.java]: Fixed in commit 166af79795 by casting GZ_BYTES.length to long before multiplication. This preserves the test behavior and prevents the potential integer overflow. The focused processing tests passed.

File: processing/src/test/java/org/apache/druid/utils/CloseableUtilsTest.java [hunk 3]
```diff
@@ -79,7 +80,7 @@
     MatcherAssert.assertThat(e, CoreMatchers.instanceOf(IOException.class));
 
     // Second exception
-    Assert.assertEquals(1, e.getSuppressed().length);
+    Assertions.assertEquals(1, e.getSuppressed().length);
```
[the remark this answers is not in the document]
Comment (FrankChen021) [reply processing/src/test/java/org/apache/druid/utils/CloseableUtilsTest.java:86]: Fixed in commit 166af79795 by capturing the expected exception with Assertions.assertThrows instead of a nullable local and manual try/catch. The focused processing tests passed.

File: processing/src/test/java/org/apache/druid/utils/CloseableUtilsTest.java [hunk 4]
```diff
@@ -108,7 +109,7 @@
...
```
[the remark this answers is not in the document]
Comment (FrankChen021) [reply processing/src/test/java/org/apache/druid/utils/CloseableUtilsTest.java:112]: Fixed in commit 166af79795 by capturing the expected exception with Assertions.assertThrows instead of a nullable local and manual try/catch. The focused processing tests passed.

File: processing/src/test/java/org/apache/druid/testing/junit/LoggerCaptureExtension.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  private static class InMemoryAppender extends AbstractAppender
+  {
+    private static final String NAME = InMemoryAppender.class.getName();
+    private final String targetLoggerName;
+    @GuardedBy("logEvents")
+    private final List<LogEvent> logEvents = new ArrayList<>();
+
+    InMemoryAppender(final Class<?> targetClass)
+    {
+      super(NAME, null, null);
```
[the remark this answers is not in the document]
Comment (FrankChen021) [reply processing/src/test/java/org/apache/druid/testing/junit/LoggerCaptureExtension.java]: Fixed in commit 3b92250d3c by replacing the deprecated three-argument AbstractAppender constructor with the supported five-argument overload and Property.EMPTY_ARRAY. Processing test compilation passed with mvn -ntp -pl processing test-compile -Dweb.console.skip=true -T1C.

File: processing/src/test/java/org/apache/druid/java/util/common/StringUtilsTest.java [hunk 6]
```diff
@@ -150,165 +145,171 @@
...
   @Test
   public void testNonStrictFormat()
   {
-    Assert.assertEquals("test%d; format", StringUtils.nonStrictFormat("test%d", "format"));
-    Assert.assertEquals("test%s%s; format", StringUtils.nonStrictFormat("test%s%s", "format"));
+    Assertions.assertEquals("test%d; format", StringUtils.nonStrictFormat("test%d", "format"));
+    Assertions.assertEquals(
+        "test%s%s; format",
+        // codeql[java/missing-format-argument]
+        StringUtils.nonStrictFormat("test%s%s", "format")
```
[the remark this answers is not in the document]
Comment (FrankChen021) [reply processing/src/test/java/org/apache/druid/java/util/common/StringUtilsTest.java:162]: Evaluated and intentionally left unchanged: this is a valid CodeQL pattern match but an expected test-only case for StringUtils.nonStrictFormat fallback behavior. The alert was dismissed as used in tests; no source behavior change is needed. The focused StringUtilsTest and processing validation passed.

File: processing/src/test/java/org/apache/druid/segment/CloserExtension.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+  public CloserExtension(final boolean throwException)
+  {
+    this.throwException = throwException;
+  }
+
+  @Override
+  public void beforeEach(final ExtensionContext context)
+  {
+    closer = Closer.create();
```
Comment (FrankChen021) [inline processing/src/test/java/org/apache/druid/segment/CloserExtension.java]: [P2] beforeEach discards early registrations

If closeLater() is called during test-instance construction, the extension registers the resource on one Closer, then unconditionally replaces it in beforeEach; the resource is never closed. Preserve an existing closer or initialize it without overwriting registrations.

File: processing/src/test/java/org/apache/druid/collections/bitmap/BitmapIterationTest.java [hunk 8]
```diff
@@ -19,135 +19,42 @@
...
-    TestSuite suite = new TestSuite();
-    for (BitmapFactory factory : factories) {
-      suite.addTest(suiteForFactory(factory));
-    }
-    return suite;
+    return List.of(new BitSetBitmapFactory(), new ConciseBitmapFactory());
   }
 
-  private static Test suiteForFactory(BitmapFactory factory)
+  @ParameterizedTest
```
Comment (FrankChen021) [inline processing/src/test/java/org/apache/druid/collections/bitmap/BitmapIterationTest.java:59]: [P2] Bitmap iterator contract coverage was removed

The former Guava suite covered arbitrary sizes, ordering, duplicates, restricted elements, and iterator contract behavior. The replacement checks only [1, 3] and empty iteration, allowing regressions in boundary and exhaustion behavior to pass. Port the contract cases to Jupiter or retain equivalent generated coverage.

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 0 |
| P2 | 2 |
| P3 | 0 |
| Total | 2 |

Reviewed 61 of 61 changed files.

The inline comments cover a test-extension lifecycle leak and materially reduced bitmap iterator contract coverage.

Validation: git diff --check passed; no Maven tests or builds were run.



---

This is an automated review by Codex GPT-5.6-Luna(max)

File: processing/src/test/java/org/apache/druid/segment/CloserExtension.java (hunk 7 above)
Comment (FrankChen021) [reply processing/src/test/java/org/apache/druid/segment/CloserExtension.java]: Fixed in commit 5b968b3f0e: CloserExtension.beforeEach now initializes the Closer only when it is null, preserving resources registered during test-instance construction. Processing validation passed; the focused BitmapIterationTest passed all 14 tests.

File: processing/src/test/java/org/apache/druid/collections/bitmap/BitmapIterationTest.java (hunk 8 above)
Comment (FrankChen021) [reply processing/src/test/java/org/apache/druid/collections/bitmap/BitmapIterationTest.java:59]: Fixed in commit 5b968b3f0e: restored parameterized Jupiter coverage for empty, singleton, multi-element, arbitrary-size, ordering, duplicate, exhaustion, and unsupported-mutation cases. BitmapIterationTest passed all 14 tests, and processing validation passed.

**Review by FrankChen021 — COMMENTED** [review]
reviewed all changes made by Codex, LGTM

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the code for correctness, edge cases, concurrency, and integration risks; no issues found.

Reviewed 61 of 61 changed files.

Validation: full and incremental git diff --check passed; no builds or tests were run.



---

This is an automated review by Codex GPT-5.6-Luna(max)

**Review by GWphua — APPROVED** [review]
